src/common/protobuf_utils.hpp

```diff
@@ -64,6 +64,63 @@ struct Slave;
 
 namespace protobuf {
 
+//
+// A "protobuf union" is a protobuf message
+// which has an enum named "Type" defined in its scope,
+// a required field named "type" of a "Type" type
+// and a field corresponding to each member of the "Type" enum.
+//
+// A valid "protobuf union" should satisfy the following requirements.
+// 1. A member of the Type enum with a number (not index!) of 0,
+//    if there is any, should be named "UNKNOWN".
+// 2. For each other member of the Type enum there should be an optional field
+//    in the message with an exactly matching name in lowercase.
+// 3. The fields that correspond to members of the Type enum other than
+//    message.type(), should not be set.
+//
+// NOTE: If possible, oneof should be used in the new messages instead
+// of the "protobuf union".
+//
+class UnionValidatorHelper
+{
+public:
+  UnionValidatorHelper(
+      const google::protobuf::Descriptor*,
+      const google::protobuf::EnumDescriptor*);
+
+  Option<Error> Validate(
+      const int messageTypeNumber, const google::protobuf::Message&) const;
+
+private:
+  std::vector<std::pair<int, const google::protobuf::FieldDescriptor*>>
+    UnionFieldDescriptors;
+  const google::protobuf::EnumDescriptor* TypeDescriptor;
+};
+
+
+template <class Message>
+class UnionValidator
+{
+public:
+  static Option<Error> Validate(const Message& message)
+  {
+    static const UnionValidatorHelper helper(
+        Message::descriptor(), Message::Type_descriptor());
+    return helper.Validate(message.type(), message);
+  };
+  UnionValidator() = delete;
+};
+
+
+// This function returns Error if the protobuf union is not valid
+// and Nothing otherwise.
```

Comment:
Let's also mention that it will abort the process if the passed message is not a protobuf union. (which is actually the correct behaviour, I think, because getting the wrong message type here is essentially a compile-time bug that is only detected at runtime)